### PR TITLE
Feature/kas 3254 opladen van een word versie van een besluit op de besluiten tab bij publicaties voor publicaties via mr

### PR DESCRIPTION
**Pull request by sergiofenoll**
Backend PR: https://github.com/kanselarij-vlaanderen/app-kaleidos/pull/260

Adds functionality to upload a .doc(x) version of a document that's part of a publication via a MR.

To keep backwards compatibility (and to stick to the provided solution in the ticket), a `get file()` and `set file()` were implemented in the `piece` model to emulate the behaviour of having only one file. The getter works by returning one of the files in the following order of priority: docx, pdf, whichever file was first in the array, or `undefined`.

DocumentPreviewModal has also been expanded to support multiple files per piece, allowing users to choose which file to render. Though unsupported file types are not renderable (and thus not selectable), so in practice users can only select PDFs.

Also added support to the FileUploader for file validation. Here it is used to only allow users to upload .doc or .docx files in the described flow.

Some screenshots added so you can see the UI at a glance.

Screenshot from 2022-03-22 09-52-44

Screenshot from 2022-03-22 09-52-48

Screenshot from 2022-03-22 09-52-58

Screenshot from 2022-03-22 09-53-11

### CODE REVIEW

**Comment by sergiofenoll** [comment]
Custom Jenkins build: http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination/168/

**Comment by sergiofenoll** [comment]
Seeing some failing tests, will try to look at each one. Starting with the first one I tackled:

http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination/168/artifact/reports/cypress_screenshots/screenshots/unit/decision.spec.js/Decision%20tests%20--%20should%20test%20the%20document%20CRUD%20for%20a%20decision%20(failed).png

Error thrown is the following:
`Assertion Failed: Cannot call toArray before content is assigned.`

I could reproduce this locally by navigating to an Agenda > Agendapunt > Beslissing > Beslissingfiche uploaden > Uploading a file then trying to delete the file.

Looking at the console, this is the stacktrace:
```
Error occurred:

- While rendering:
  -top-level
    application
      agenda
        agenda.agendaitems
          agenda.agendaitems.agendaitem
            agenda.agendaitems.agendaitem.decisions
              agenda/agendaitem/agendaitem-decision
                utils/add-document
                  web-components/vl-modal
                    click-outside
                      auk/modal/body
                        ember-kaleidos-webuniversum/vl-uploaded-document

Uncaught (in promise) Error: Assertion Failed: Cannot call toArray before content is assigned.
    Ember 2
    get file piece.js:114
    Ember 2
    child reference.js:222
    tag reference.js:170
    track validator.js:835
    valueForRef reference.js:169
    child reference.js:219
    tag reference.js:170
    track validator.js:835
    valueForRef reference.js:169
    ref reference.js:261
    tag reference.js:170
    track validator.js:835
    valueForRef reference.js:169
    Ember 2
    evaluate runtime.js:3863
    _execute runtime.js:5285
    execute runtime.js:5255
    runInTrackingTransaction validator.js:165
    execute runtime.js:5255
    rerender runtime.js:5634
    Ember 3
    inTransaction runtime.js:5074
    Ember 3
    invoke backburner.js:351
    flush backburner.js:241
    flush backburner.js:447
    _end backburner.js:999
    end backburner.js:729
    _run backburner.js:1054
    run backburner.js:773
    Ember 3
    _run backburner.js:1052
    _join backburner.js:1028
    join backburner.js:779
    Ember 3
    deletePieceInCreation add-document.js:96
    Ember 6
    _run backburner.js:1052
    _join backburner.js:1028
    join backburner.js:779
    Ember 3
    jQuery 8
    Ember 5
    invoke backburner.js:351
    flush backburner.js:241
    flush backburner.js:447
    _end backburner.js:999
    end backburner.js:729
    _run backburner.js:1054
    run backburner.js:773
    Ember 8
    create index.js:64
    Ember 2
```

So in the `Utils::AddDocument` component, when deleting the uploaded file we're calling this method: https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/6f07b413f8200f375d9bd03b00c52113c7704115/app/pods/components/utils/add-document/component.js#L31

The fix I found locally was by first saving the piece when uploading a file, and calling `destroyRecord()` instead of `deleteRecord`() in the above method. Right now, this method is working on a piece that has been created (call to `store.createRecord('piece', {})`) but not saved. If I add a call to `this.pieceInCreation.save()` after creating it, then destroying the records does not cause an exception.

Although ideally this would work without having to persist the records first.

**Comment by sergiofenoll** [comment]
http://kal-kastaar.s.redpencil.io:8080/job/kaleidos/job/custom_frontend_backend_combination/168/artifact/reports/cypress_screenshots/screenshots/unit/file-upload.spec.js/Add%20files%20to%20an%20agenda%20--%20should%20delete%20documents,%20pieces%20and%20files%20(failed).png

Error:
``Attempted to handle event `pushedData` on <piece:5EBA8CF9DAB6BB000900000D> while in state root.deleted.inFlight.``

Stacktrace in console after reproducing:
```
Uncaught (in promise) Error: Attempted to handle event `pushedData` on <file:623a1fa89566bc000d00001e> while in state root.deleted.inFlight. 
    Ember 6
    _run backburner.js:1052
    _join backburner.js:1028
    join backburner.js:779
    Ember 2
    invokeCallback rsvp.js:495
    publish rsvp.js:478
    <anonymous> Ember
    invoke backburner.js:351
    flush backburner.js:241
    flush backburner.js:447
    _end backburner.js:999
    _boundAutorunEnd backburner.js:648
    promise callback*buildNext/< backburner.js:28
    flush Ember
    _scheduleAutorun backburner.js:1218
    _end backburner.js:1009
    _boundAutorunEnd backburner.js:648
    promise callback*buildNext/< backburner.js:28
    flush Ember
    _scheduleAutorun backburner.js:1218
    _ensureInstance backburner.js:1206
    schedule backburner.js:805
    <anonymous> Ember
    fulfill rsvp.js:430
    handleMaybeThenable rsvp.js:386
    resolve$1 rsvp.js:403
    error rsvp.js:343
    promise callback*tryThen rsvp.js:324
    handleForeignThenable rsvp.js:333
    <anonymous> Ember
    invoke backburner.js:351
    flush backburner.js:241
    flush backburner.js:447
    _end backburner.js:999
    _boundAutorunEnd backburner.js:648
    promise callback*buildNext/< backburner.js:28
    flush Ember
    _scheduleAutorun backburner.js:1218
    _end backburner.js:1009
    _boundAutorunEnd backburner.js:648
    promise callback*buildNext/< backburner.js:28
    flush Ember
    _scheduleAutorun backburner.js:1218
    _end backburner.js:1009
    _boundAutorunEnd backburner.js:648
    promise callback*buildNext/< backburner.js:28
    flush Ember
    _scheduleAutorun backburner.js:1218
    _end backburner.js:1009
    _boundAutorunEnd backburner.js:648
    promise callback*buildNext/< backburner.js:28
    flush Ember
    _scheduleAutorun backburner.js:1218
    _ensureInstance backburner.js:1206
    schedule backburner.js:805
    <anonymous> Ember
    fulfill rsvp.js:430
    handleMaybeThenable rsvp.js:386
    resolve$1 rsvp.js:403
    error rsvp.js:343
    promise callback*tryThen rsvp.js:324
    handleForeignThenable rsvp.js:333
    <anonymous> Ember
    invoke backburner.js:351
    flush backburner.js:241
-private.js:6474
    _end backburner.js:1020
    _boundAutorunEnd backburner.js:648
    (Async: promise callback)
    buildNext backburner.js:28
    flush Ember
    _scheduleAutorun backburner.js:1218
    _end backburner.js:1009
    _boundAutorunEnd backburner.js:648
    (Async: promise callback)
    buildNext backburner.js:28
    flush Ember
    _scheduleAutorun backburner.js:1218
    _end backburner.js:1009
    _boundAutorunEnd backburner.js:648
    (Async: promise callback)
    buildNext backburner.js:28
    flush Ember
    _scheduleAutorun backburner.js:1218
    _end backburner.js:1009
    _boundAutorunEnd backburner.js:648
    (Async: promise callback)
    buildNext backburner.js:28
    flush Ember
    _scheduleAutorun backburner.js:1218
    _end backburner.js:1009
    _boundAutorunEnd backburner.js:648
    (Async: promise callback)
    buildNext backburner.js:28
    flush Ember
    _scheduleAutorun backburner.js:1218
    _end backburner.js:1009
    _boundAutorunEnd backburner.js:648
    (Async: promise callback)
    buildNext backburner.js:28
    flush Ember
    _scheduleAutorun backburner.js:1218
    _end backburner.js:1009
    _boundAutorunEnd backburner.js:648
    (Async: promise callback)
    buildNext backburner.js:28
    flush Ember
    _scheduleAutorun backburner.js:1218
    _ensureInstance backburner.js:1206
    schedule backburner.js:805
    <anonymous> Ember
    fulfill rsvp.js:430
    handleMaybeThenable rsvp.js:386
    resolve$1 rsvp.js:403
    error rsvp.js:343
    (Async: promise callback)
    tryThen rsvp.js:324
    handleForeignThenable rsvp.js:333
    <anonymous> Ember
    invoke backburner.js:351
    flush backburner.js:241
```

Couldn't find a fix for this locally (yet)

File: app/components/publications/documents/document-card-step.js [hunk 1]
```diff
@@ -1,7 +1,44 @@
...
+ * @argument {Piece} piece
+ * @argument {boolean} isEnabledDelete
+ * @argument {boolean} isViaCouncilOfMinisters
+ * @argument {Function} onDelete
+ */
 export default class PublicationsDocumentsDocumentCardStepComponent extends Component {
+  @tracked isOpenWordDocumentUploadModal;
+
+  get hasWordVersionOfDocument() {
+    const file = this.args.piece.file;
```
Comment (sergiofenoll) [inline app/components/publications/documents/document-card-step.js:16]: Await this relation in a task, and don't show the layout until the task has been finished (so this getter always has `this.args.piece.file`)

**Comment by erikap** [comment]
Story has more impact than expected. Closing this PR for now. Document flow/representations will be revisited and majorly refactored in the (hopefully near) future.

**Pull request closed without merging**